File: Caelan.Frameworks.Common/Helpers/PasswordHelper.cs
using System;
using System.Linq;
using System.Security.Cryptography;
using System.Text;

namespace Caelan.Frameworks.Common.Helpers
{
    public abstract class PasswordHelper
    {
        public abstract string GetSalt();
        protected abstract string GetDefaultPassword();

        public virtual string GetDefaultPasswordEncrypted()
        {
            return EncryptPassword(GetDefaultPassword());
        }

        private static string Sha512Encrypt(string password)
        {
            return String.Join("", (new SHA512CryptoServiceProvider()).ComputeHash(Encoding.Default.GetBytes(password)).Select(b => b.ToString("x2").ToLower()));
        }

        public virtual string EncryptPassword(string password)
        {
            return Sha512Encrypt(GetSalt() + Sha512Encrypt(password));
        }
    }
}

File: Caelan.Frameworks.Common/Helpers/PasswordHelper.fs
namespace Caelan.Frameworks.Common.Helpers

open System.Security.Cryptography
open System.Text
open Caelan.Frameworks.Common.Interfaces

type PasswordHelper(salt : string, defaultPassword : string, encryptor : IPasswordEncryptor) = 
    member this.GetSalt() = salt
    member this.GetDefaultPassword() = defaultPassword
    member this.GetDefaultPasswordEncrypted() = this.EncryptPassword(this.GetDefaultPassword())
    member this.EncryptPassword(password) = 
        password |> MemoizeHelper.Memoize(fun p -> encryptor.EncryptPassword(this.GetSalt() + encryptor.EncryptPassword(p)))
    new(salt, defaultPassword) = 
        let encryptor = 
            { new IPasswordEncryptor with
                  member __.EncryptPassword(password) = 
                      using (new SHA512CryptoServiceProvider()) (fun provider -> 
                          provider.ComputeHash(Encoding.Default.GetBytes(password))
                          |> Array.map (fun t -> t.ToString("x2").ToLower())
                          |> String.concat "") }
        PasswordHelper(salt, defaultPassword, encryptor)
